### ISSUE TITLE
Finishing 4.0

### ISSUE THREAD

**Issue by TheSnacker** [eternaldensity/Sandcastle-Builder#1385]
Admittedly, I have been kinda missing from all of this for a while.  Sorry about that.  It is kinda hard to get myself to program after working all day.  I also got kinda intimidated when it came time to start making new stuff and not just fix what was already there.  That said, I am going to try to pitch in and help fix what needs it.  Given that there hasn't been an update in over two weeks, it sounds like it is needed.  I do need to know what needs fixing though.  I have tried looking through the other topics and all I have found is one item.  Please reply with any others that still need fixing, and I'll try to add them to this list and work on them.

https://docs.google.com/spreadsheets/d/1ry0g-n26zvJiuPVQr4AnTlcY8B2F3_6FTAUa5sYloaA/edit#gid=493290078

edit: replaced list with speadsheet link

**Comment by pickten** [comment]
Dunno if what is fixed vs what is wrong on the spreadsheet, but, on the spreadsheet:

 Breath still seems to lock on use.

Knight attacks become too hard too fast

Glass lightning notif spam

Also, the overview is listed as fixed on the spreadsheet.

**Comment by TheSnacker** [comment]
Okay, that helps.  I didn't find the spreadsheet before.  I have added a link to it in the main post here.  It isn't the most straightforward thing, but I will see if I can figure out what still needs to be done through it and see what I can do to fix things.

**Comment by eternaldensity** [comment]
Cool, I was actually thinking of sending out a reminder that this still
exists :D

**Comment by EPSIL0N** [comment]
I think the breath fix is waiting to be merged in, I am going to look into the knight fight issues.

**Comment by eternaldensity** [comment]
There's no open pull requests on my end.

**Comment by EPSIL0N** [comment]
Ah, I think there is a request to merge into LL's branches that I will probably end up completing this weekend if not tonight.

Not trying to take over things but just because few other people change the issues when you look at the colours of the issues they generally reflect my opinion on the issue, green being solved, yellow for minor stuff or just needing some testing and red/orange for more major concerns.

I think we were are aiming for all green.

**Comment by TheSnacker** [comment]
Sorry, I tried to work on this over the weekend, but I had no idea how far thing have come, and I'm a little lost at this point.  I'm thinking I'm going to switch over to the beta version soon, and start working my way through just so that I can see what is actually being talked about so that I may even be able to fix something.

**Comment by EPSIL0N** [comment]
Ooh, another person. I think I have the absolute most up to date stuff on my local copy as I seem to have fallen into the position of the person who merges all the different collaborators work along with about two weeks of bug fixes. I could push it to somewhere on github if you wanted.

**Comment by TheSnacker** [comment]
I was going to start playing through the one linked in the "Tired of this" thread in the Reddit.  If that is not the most up to date version though, it would help to have access to that.

I still hurt for time to actually program, but I can QA at least, since I can play the game at work without it interfering too much.

**Comment by EPSIL0N** [comment]
Ah, what is linked is not too out of date but I wish I had pushed an update a week ago as pretty much as soon as I offered to update last night we found that one of the fixes has created an incorrect save loading behavior, hopefully it won't be too hard to work it out.

**Comment by TheSnacker** [comment]
Running through that version, I did find another bug.  It looks like Robotic Hatcher is not working.  I added it to the doc.

**Comment by EPSIL0N** [comment]
Status update. I am sitting on a lot of fixes and tweaks but atleast one of them changes the way saves are read which causes issues for the beta testers as it makes current version 4 not load cleanly into possible upcoming version 4.

On that note I wanted to confirm that we are aiming for version 4 to be the number on release and to ask if we want beta saves to be carried over or not. I like the idea of a flat 4 and have no strong feelings either way. I would like to keep my progress and I took steps to avoid using any thing we found as broken to get an advantage.

We are currently waiting on a few more tests:
Does the WotT/WotP one or the other system work properly?
Does aperture science cause a loot page overflow with 1000 copies / Does that only happen before save and reload?
Does Blue Hints effect both per click and per mnp like the code suggests it should?

And the remaining unclaimed issures on the todo list are:
Make kitten reward system for dragon drum and sea mining and leo.
Add notifs and effects to those three as needed.
Green Sun needs a unlock condition.
Honor among serpents notification cancel bug appears to not be fixed,
once fixed Honor among serpents will need an unlock condition.


The Honor among serpents bug looks pretty simple so I will likely move on to that shortly.
I believe pickten is working on a few issues and I am trying to help.

**Comment by TheSnacker** [comment]
I tried running through the new stuff, but there is so much that is new, and this being sandcastle builder, it is hard to figure everything out on my own. I could dig through the code to see what the new boosts are, and the unlock conditions or whatnot, but that would take a lot of time.  I'd ask you guys to help me through it, but at this point, I wouldn't be helping as much as I would be holding you guys back.

I came in here thinking that it would just be squashing bugs, and that I was good at that.  Sadly I would need to learn the new game before I could do any of that, and by the time I learned that, I'm sure you guys will have the bugs beat.  For now, I'm going to go back to the game as it is, and try to get infinite bonemeal, because at the point I was at in 4.0 when I stopped, though the diamonds were there, winning all the dragon fights that I would need to get all the GDLP unlocks would have been next to impossible.

**Comment by EPSIL0N** [comment]
Ah, thanks for the help, every extra viewpoint moves us forward even if it is just by making me a little more confident I've not been over looking some behavior for an entire month.

I'm not sure if it is meant to be possible to get to infinite bonemeal, I'm trying to figure out what it supposedly breaks besides infinite bonemeal to infinite eggs which is not a good idea.

Yeah the knight attack fights in 4 are pretty broken because of some code involving princesses but that is one of the things I have a change for. I think it is something like the fight after Lords is a sextillion times harder and the next one like an octillion times that?

**Comment by TheSnacker** [comment]
The problem is that by the time I stopped, the best I could do was the early Lord fights.  Even without princesses pushing it past the Lords stage, Wyverns get attacked by late Lords by default.  Princesses break it worse, but it kinda seems to be broken even before that, unless I am missing something?

**Comment by EPSIL0N** [comment]
I see where you are coming from. I think I just ignored knights when I was playing and I also changed that a little while doing the rework. I think I could only reliably beat 2 lords but I haven't taken any of the specializing super boosts yet.

**Comment by TheSnacker** [comment]
Oddball question.  I notice that the cats are now devoid of chromatic heresy, and I was wondering if it was just that, to make it black and white like everything else, or if there was more to it?

**Comment by pickten** [comment]
What it was was that I don't know the colored variant's address, and didn't see it anywhere in the existing code. Does anyone know?

**Comment by TheSnacker** [comment]
Oh.  Looking through the code, you just need to take out the /g to get color versions.  It looks like this is set up in gui.js?v=1.

**Comment by EPSIL0N** [comment]
We are still plugging away at it, I didn't make much progress this last week due to real life stuff.

We still need suggestions for unlock conditions for Green Sun and Honor among serpents.

We also need clarification if people in general or the authors feel that Dragon Drum and Leo should be available the moment they can have effect or require additional conditions be met.

I was also looking for opinions if the kitten only rewards should be effected by double department as twice as many could be nothing of note like 2 panes instead of 1 or it could be pretty big for doubled dragon drum. I don't think Sea Mining is effected by doubling.

**Comment by Calamitizer** [comment]
Hey, so. Sorry I haven't been around in forever, my computer sort of... died. It won't charge, and I haven't been able to get a new one in a few months, so no programming for me. I've been following along on mobile but I'm not sure what I can contribute. Regarding LeoLeo, though, the idea is that itkup

**Comment by pickten** [comment]
I feel like Green Sun should probably unlock when you get your Nth Dimensional Keyhole for some N, probably one of 2, 5, or 11 (first t1i discovery).

Honor Among Serpents seems weak to me, though possibly because I'm only a 'newt player, so maybe just a wyvern fight reward? Alternately, maybe just make it be when you fledge on top of hiding wyverns? IDK, it just doesn't seem like it needs to be any harder than that.

For the Double Department thing, I think it would be a bit weird for DD to affect the kitten only rewards, since the kitties are separate from the DoRD in theory, but I don't know how the balance would work out.

**Comment by EPSIL0N** [comment]
Ah good to hear from you Cal.

I think that is a good idea related to some milestone in Dimensional keys or progress in t1i (I had missed the note in the code that says that)

Yeah, I think if we've managed to fix the bug with Honor it is no longer overpowered.

I think the kitten issue was mainly in my head, it should be pretty easy to have 1 kitten reward or if none are available than 1 or 2 depending on department DoRD rewards.

**Comment by EPSIL0N** [comment]
Aiming to try and get something together for this weekend if not Friday but I was wondering how many people here are interested in testing more, the save issue I might have mentioned before still exists but it is an isolated case to the testers so it is hard to estimate if I should put time into writing a persist clause for it or just offer to convert the saves for the few people who will be effected.

**Comment by pickten** [comment]
I'd say we should just write a separate script, but I'm not sure how many people would use it/need it.

**Comment by EPSIL0N** [comment]
Progress report. Just about ready to push the update, ended up having to alter persist anyways so I've gone ahead and done the relatively small apparently changes to load testing 4 saves into 4.001 (or whatever we want to call it that is >4). Just hunting a bug that we need to deal with as it will persist into the future and be a relative annoyance to all new players and doing some final testing.

**Comment by Calamitizer** [comment]
How's shard/pane stuff doing? What are the outstanding tasks (if any)? I've got no idea what's happened since my computer gave up. Has it actually been playtested?

**Comment by EPSIL0N** [comment]
The update as a whole has not really been play tested but I think I have tested each of the smaller fixes as I merge or write them into my local copy.

I believe everything shard/pane related is in working order but there may be balance issues. Green sun gets pretty crazy powerful within about two days of t1i progress.

**Comment by EPSIL0N** [comment]
Looks to be ready for the next round of testing, just going to give people a little bit longer to back up their saves. I think the issue is fixed but if I tested with ten different saves, the eleventh could still break.

**Comment by EPSIL0N** [comment]
Good news is I had the foresight/good fortune to push the month's worth of content from my laptop to github, bad news is an hour later my laptop battery then reached the end of the 'end of life' stage and apparently this is not a model that tolerates 'AC only' running well. I'm not completely out but my next best choice is a desktop so I won't be able to check the spreadsheet or github and reddit threads as spread out during the day on weekdays.

**Comment by erp-lsf** [comment]
Greetings.
inb4: sorry, for jumping right into the testing, and maybe, not reading previous discussion.
I've imported my savefile into last version of Leopards master, and noted that I can't Time Travel, or Negator-flip into negative NPs. Is it known bug, or intended behaviour?

**Comment by pickten** [comment]
I know what's wrong with Negator -- that's from a check I added, not realizing Negator uses the same function as time travel. What do you mean for Time Travel, though? What NP were you trying to travel from and to?

**Comment by EPSIL0N** [comment]
Additional details like which NP to which NP would help a lot as I may have broken something while trying to fix something else.

**Comment by erp-lsf** [comment]
Here's save file for the case study:
http://sprunge.us/eHNF
I'm trying to travel backwards via Time Travel from NP 1 and recieve 'Divide by zero error'. Jumping via discovery to NP -1 puts me into flipped NP 0, traveling forward via TT from NP -1 does the same. Using Camera on that NP reports that I already have the discovery for 'Minus In The Beginning'.
Then, trying to travel from flipped NP 0 forward also reports divide by zero error.

**Comment by EPSIL0N** [comment]
Ah, Thank you. I'll try to look at it some time today, I wonder what the issue is this time.

**Comment by EPSIL0N** [comment]
Found it and fixed it, it was just a visual bug but I also added in pickten's changes to my working copy, so I am wondering when I should push an update?

**Comment by erp-lsf** [comment]
I think you can push updates into LL's repository, whenever you're done.

**Comment by EPSIL0N** [comment]
There are a few little things on my list to address which I will try to do in the next few days but most of what is not quite satisfactory to me is the stuff that has been sitting in opinion limbo for weeks now.

Unless someone comes up with a larger complaint or issue I think the only thing left to consider is when we would want to release it in terms of people being available to fix any major bugs or balance issues that appear with the bigger audience.

**Comment by eternaldensity** [comment]
I was thinking of asking how things were going, and I see that you got
there first :D
Yeah it's about time we got something out the door soon.  So long as
there's nothing game breaking or save breaking or really really dumb...
(more than the norm, I mean. ;) )

**Comment by erp-lsf** [comment]
I'm eager to all new stuff that's coming, and a few bugs won't matter so much :)

**Comment by EPSIL0N** [comment]
I have pushed the version with the last of the things from my todo list and a few more typo fixes from pickten to my EPS-fixes branch.

I don't know of anything that should break but some of the dragon fight mechanics relating to princesses are pretty much placeholders as the fights went from completely unwinnable to currently a touch too easy.

**Comment by pickten** [comment]
There seems to be something weird causing the prey.indexof bug that occasionally shows -- I had it briefly, refreshed, and now it's gone.

Also, because it's bugging me, I'm trying to get the negpix overview to actually display negpix info (wasn't an issue before because it was unfledgeable anyways).

**Comment by EPSIL0N** [comment]
I didn't think we were at the stage to fledged in the negatives yet.

**Comment by pickten** [comment]
Oh, in that case the message should be much clearer (it says "0 __ have been fledged at NP__"). I have been focusing more on the display issues, so I'll just adjust the message in the case of negative newpixes to say something along the lines of "you can't do that" and leave it at that.

**Comment by TheSnacker** [comment]
Change it to: "The underworld is no place for living dragons"

I have an idea for using negative NPs as a place to fledge undead dragons to, using a Necromancer boost.  I think it would be an interesting idea for a use for Dragons lost fighting and Starved Dragons, as well as an interesting use for Negative pix.  I do not want it to be something that 4.0 is delayed any further for though.  Far too many people have waited far too long for this update.  We should not make them wait any longer if they do not have to.

**Comment by eternaldensity** [comment]
*nods*

**Comment by pickten** [comment]
That makes sense, and I'm very curious to see how that turns out, so I'll change that shortly.

**Comment by elswyyr** [comment]
What's missing now? Is there anything that still needs to be done before the update is pushed to live?

**Comment by eternaldensity** [comment]
I was gonna ask the same.  If I went live today, what terrible things woukd
result?

**Comment by pickten** [comment]
There was the bug identified on the sub a couple days back allowing unlimited negative downgrading of the refinery/chiller that I want to check/remove, but haven't gotten to yet (I'll probably get it done tonight, though). While not a bug, per se, there is also something that seems to cause button-presses to scroll back up to the top of the page. Besides that, I think the only known bug is the daylight savings time ONGs, but that can easily be pushed back.

**Comment by EPSIL0N** [comment]
I can't think of any thing else that is broken.

On the note of the dst ONGs I have no idea where to even start with it as I can't seem to make it happen for me.

**Comment by pickten** [comment]
For DSTs, I'm not sure it's happening either, but I'm going off of what has been reported on the spreadsheet. I'll try to look into it tonight or tomorrow, mostly. I'll try to see if it's happening tonight. Unfortunately, my `Date` modifications are not working yet. I'll edit with a yay/nay when I get them working.

**Comment by eternaldensity** [comment]
Remind me, is DST ONGs an in-game thing, or do you mean real-world DST
changes are breaking ONGs?  Actually I vaguely remember hearing something
about the latter.

Does the game just lock up for an hour when DST drops back an hour in
time?  Cos that would make sense given the anit- "mess with the system
clock as a way to cause ONGs" code I remember writing.
Mind you, people can achieve the same thing by just running Molpy.ONG so
maybe that protection is silly and not needed and causes more problems than
it hypothetically fixes.

**Comment by pickten** [comment]
When DST shifted around November 1st, everyone whose computer was running SB got 1 ONG per mNP for around 2000mNP, which is problematic because it's just so many and was very confusing to people unaware of the bug. At the very least, I think it would be good to have some sort of explanatory notif about it when it strikes or maybe a minor boost.

**Comment by eternaldensity** [comment]
Okay that's severely extreme. ]:(

**Comment by eternaldensity** [comment]
So the actual version we want to release with is 4.0, not 4.001, correct?

**Comment by eternaldensity** [comment]
> When DST shifted around November 1st, everyone whose computer was running SB got 1 ONG per mNP for around 2000mNP, which is problematic because it's just so many and was very confusing to people unaware of the bug. At the very least, I think it would be good to have some sort of explanatory notif about it when it strikes or maybe a minor boost.

I assume by 'everyone' you mean everyone who had a DST shift.  And I assume you mean the northern hemisphere's DST, so that would be DST ending which means the time becomes an hour... earlier?

**Comment by eternaldensity** [comment]
Welp.

I was testing just fine and I managed to replicate the error and was working on figuring out the exact cause (it's in ONGsnip, which sets the minutes to 00 or 30 without caring about stuff like DST) and everything was going fine until I refreshed the page and jQuery stopped working because of 'insecure response' from ajax.googleapis.  And then I found Google wasn't working either, because it couldn't confirm it's ssl certificate.

Then I realised that my adjustments to my system clock had taken me outside the range of the valid time of google's ssl certificates.  Whoops!

Oh and I got logged out of github too :P

**Comment by pickten** [comment]
Re: the first. I should think so. It wouldn't make much sense and the version was only booted to 4.001 for automatic save-fixing for testing purposes, which doesn't/shouldn't apply to the live version.

Re: the second comment. I don't know much about DST beyond US DST and that Europe had a DST shift a week or two prior, so I guess.

Re: the third. Nice. Because of shenanigans like that, I tried a pure js approach (redefining `Date` functions, mostly), but that either crashed or simply didn't do anything. TBH I was considering a stronger approach that checks for signs of DST (namely <-1 mNP left to ong) and manually fixes it, which might be easier to code but would be less future-proof. I'm not sure what would happen when DST ends, though, which is another downside to that approach. Knowing that it comes from ONGsnip makes that pretty irrelevant, though.

**Comment by eternaldensity** [comment]
I'm rewriting the time stuff using moment.js which should make the time manipulation stuff much easier.

**Comment by pickten** [comment]
Oh, that could work. I have a probable fix though which I'm about to test (using setUTC__ in ONGsnip). I'll edit once I know.

**Comment by eternaldensity** [comment]
The problem I'm working on fixing is that in the first hour after DST ends, when it tries to find the beginning of the current ONG it instead finds the time 1 hour before that.  Does your probably fix cover this case?

Using moment.js hasn't helped because it has the same bug https://github.com/moment/moment/issues/2749

But implementing it this way is just cleaner so I'll stick with it and come up with a fix.

**Comment by pickten** [comment]
No, it doesn't, because I was mostly considering when DST begins. That's a good point, but easy to fix (just mod out by ong length). I'll edit when it's fixed.

**Comment by eternaldensity** [comment]
I'm still running in circles with my method, so I expect we'll go with yours.

**Comment by eternaldensity** [comment]
I think I've got this way working alright now, but we can compare it to yours and see if I missed anything.  It turns out that I had to brute force it exactly the same time with moment.js as I would have without it, so... *shrug*

**Comment by eternaldensity** [comment]
I'mma gonna go live now!

**Comment by eternaldensity** [comment]
> Re: the first. I should think so. It wouldn't make much sense and the version was only booted to 4.001 for automatic save-fixing for testing purposes, which doesn't/shouldn't apply to the live version.

Welp.  Apparently not :P

**Comment by EPSIL0N** [comment]
Did anyone touch the dragons or the save system after I last did?
I have no clue why a bunch of dragons would be disappearing as I am fairly sure I tested it with new and old saves.

**Comment by eternaldensity** [comment]
> Did anyone touch the dragons or the save system after I last did?
> I have no clue why a bunch of dragons would be disappearing as I am fairly sure I tested it with new and old saves.

Didn't you see the resolution in the other issue?  Some genius made the dragon save system only work properly if at version 4.001, not 4.0

**Comment by EPSIL0N** [comment]
That hurts my feelings a touch. I didn't deliberately sabotage the code, I'm surprised no one else mentioned that there was a check in there for version<4.001 which at the time of testing was our working version number.

**Comment by eternaldensity** [comment]
Sorry I did't actually look to see who wrote that check.

**Comment by pickten** [comment]
Yeah, I should have remembered that that check was there, TBH, especially since I wound up doing a quick scan anyway when trying my own DST fix. Just completely slipped my mind. Looking back, it would probably have been simpler to use automated script instead.
